src/pages/library/datasetBuilder/CohortEditor.ts

```diff
@@ -0,0 +1,442 @@
+                div({ style: { marginTop: '1rem', display: 'flex', alignItems: 'center' } }, [
+                  div(
+                    {
+                      style: {
+                        height: 45,
+                        width: 45,
+                        backgroundColor: 'white',
+                        display: 'flex',
+                        alignItems: 'center',
+                        justifyContent: 'center',
+                        borderRadius: '50%',
+                        border: `1px solid ${colors.dark(0.35)}`,
+                        fontStyle: 'italic',
+                      },
+                    },
+                    ['And']
+                  ),
+                  div({
+                    style: { marginLeft: narrowMargin, flexGrow: 1, borderTop: `1px solid ${colors.dark(0.35)}` },
+                  }),
+                ]),
+              ]),
+            Utils.toIndexPairs(cohort.criteriaGroups)
+          ),
+        ]),
+  ]);
+};
+
+const editorBackgroundColor = colors.light(0.7);
+
+type CohortEditorContentsProps = {
+  updateCohort: CohortUpdater;
+  cohort: Cohort;
+  datasetDetails: DatasetResponse;
+  onStateChange: OnStateChangeHandler;
+};
+const CohortEditorContents: React.FC<CohortEditorContentsProps> = (props) => {
+  const { updateCohort, cohort, datasetDetails, onStateChange } = props;
+  return div(
+    {
+      style: { padding: `${PAGE_PADDING_HEIGHT}rem ${PAGE_PADDING_WIDTH}rem` },
+    },
+    [
+      h2({ style: { display: 'flex', alignItems: 'center' } }, [
+        h(
+          Link,
+          {
+            onClick: () => {
+              onStateChange(new HomepageState());
+            },
+            'aria-label': 'cancel',
+          },
+          [icon('left-circle-filled', { size: 32 })]
+        ),
+        div({ style: { marginLeft: 15 } }, [cohort.name]),
+      ]),
+      h3(['To be included in the cohort, participants...']),
+      div({ style: { display: 'flow' } }, [
+        h(CohortGroups, {
+          key: cohort.name,
+          datasetDetails,
+          cohort,
+          updateCohort,
+        }),
+        h(
+          ButtonOutline,
+          {
+            style: { marginTop: wideMargin },
+            onClick: (e) => {
+              updateCohort(_.set(`criteriaGroups.${cohort.criteriaGroups.length}`, newCriteriaGroup()));
+              // Lose button focus, since button moves out from under the user's cursor.
+              e.currentTarget.blur();
+            },
+          },
+          ['Add group']
+        ),
+      ]),
+    ]
+  );
+};
+
+interface CohortEditorProps {
+  onStateChange: OnStateChangeHandler;
+  datasetDetails: DatasetResponse;
+  originalCohort: Cohort;
+}
+
+type CohortUpdater = (updater: (cohort: Cohort) => Cohort) => void;
+
+export const CohortEditor: React.FC<CohortEditorProps> = (props) => {
+  const { onStateChange, datasetDetails, originalCohort } = props;
+  const [cohort, setCohort] = useState(originalCohort);
```

Comment:
useState can take a type to add very beneficial type safety:

useState<Cohort>(originalCohort);

Reply:
yes, it helps the type flow to the rest of your component code where useState is used.